pulp-glue/pulp_glue/common/authentication.py

```diff
@@ -0,0 +1,94 @@
+import json
+import typing as t
+from datetime import datetime, timedelta
+from pathlib import Path
+
+import click
+import requests
+
+
+class OAuth2ClientCredentialsAuth(requests.auth.AuthBase):
+    def __init__(
+        self,
+        client_id: str = "",
+        client_secret: str = "",
+        token_url: str = "",
+        scopes: t.List[str] = [],
+        *args: t.List[t.Any],
+        **kwargs: t.Dict[t.Any, t.Any],
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token_url = token_url
+        self.scopes = scopes
+        self.token: t.Dict[t.Any, t.Any] = {}
+
+    def __call__(self, request: requests.PreparedRequest) -> requests.PreparedRequest:
+        self.retrieve_local_token()
+
+        access_token = self.token.get("access_token")
+        request.headers["Authorization"] = f"Bearer {access_token}"
+
+        request.register_hook("response", self.handle401)  # type: ignore
+
+        return request
+
+    def handle401(
+        self, response: requests.Response, **kwargs: t.Dict[t.Any, t.Any]
+    ) -> requests.Response:
+        if response.status_code != 401:
+            return response
+
+        self.retrieve_local_token()
+        if self.is_token_expired():
+            self.retrieve_token()
```

Comment:
Talking to Grant we established that at this point it's safe to just unhook the hook from the copied request.

Reply:
Yeah, but here's the thing. The first request will always fail, since we don't have the token yet.
Then we will request one on the `handle401` method. 
Make sense to just call it once, and if it doesn't work we just finish it and call it a day?

Reply:
Done